docs/jit/ir.md

```diff
@@ -0,0 +1,257 @@
+  constructed and linked into the template during expression tree
+  construction.
+
+It is possible to create your own substitutions using the *let:*
+statement:
+
+    (template: sp_p6oget_o
+      (let:
+         (($val (load (add (^p6obody $1) $2) ptr_sz))
+         (if (nz $val) $val (vmnull))))
+
+The first subtree of the 'let:' statement is the *definition list*. A
+single *definition* consists of the *substitution name* and the
+template list that it defines. Definition lists are evaluated in
+left-to-right order, meaning that textually later definitions can
+refer to earlier definition ames. An important side effect of 'let:'
+is that *let: guarantees the evaluation order of defintions*. That is,
+the tree node represented by '$val' is computed before its reference
+in the 'if' expression following it. (This is not the case for operand
+nodes). Finally, note that 'let:' statements declare definitions in a
+single expression-wide scope and that redefinitions are not allowed.
+
+
+**Statement Macro's** are lists of which the node name start with an
+  ampersand '&'. These lists are compiled to C-macro
+  invocations. Arguments to the macro (the other elements in the list)
+  are reproduced verbatim, and nested lists are NOT supported. For
+  example:
+
+    (&MACRO_NAME foo bar 3)
+
+  Is translated to:
+
+    MACRO_NAME(foo, bar, 3)
+
+  The most common use of this is actually to use the sizeof() and
+  offsetof() expressions in field access, but it can also be used
+  to generate a constant pointer or a message string, provided you use
+  the appropriate C macro hackery.
+
+**Tree Macro's** are a facility implemented by the template
+  preprocessor to aid in developing templates. Tree macro's are
+  indicated by a '^' prefix. For instance, the '(^p6obody $1)' list
+  above is the invocation of a tree macro. A tree macro is substituted
+  directly into the invoking expression while the template is being
+  compiled. They differ from substitutions in two respects:
+
+* substitutions are local (only valid within the scope of a let:
+  expression), while macro's are globally defined. Thus, macro's can
+  be used as a building block for templates, and substitutions can
+  only be used within a template.
+
+* substitutions are *linked* into the resulting tree, thus two
+  references of the same substitution refer to the same IR tree
+  nodes. Tree macro's are *spliced* into the tree, so two
+  invocations refer to two different trees. (These trees may be
+  resolved to be equal when using common subexpression evaluation, but
+  that's a separate issue).
+
+A tree macro is defined using the 'macro:' keyword, followed by the
+macro name (including the '^' prefix), a list of macro arguments, and
+the actual macro list. For example, the '^p6obody' macro is
+implemented as:
+
+    (macro: ^spesh_slot (,a)
+       (idx
+          (load (addr (frame) (&offsetof MVMFrame effective_spesh_slots)))
```

Comment:
Also, `effective_spesh_slots` went away iirc, so this example is a tad out-dated.

Reply:
effective_spesh_slots still exists in `src/core/frame.h`, should it not?

Reply:
Oh, right, that one had to survive my MVMFrame shrinkage. :-)